demos/speech_recognition_demo/python/utils/ctc_decoder_seq_pipeline.py

```diff
@@ -0,0 +1,62 @@
+#
+# Copyright (C) 2019-2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+from copy import deepcopy
+
+from ctcdecode_numpy import SeqCtcLmDecoder
+
+from utils.alphabet import get_default_alphabet, load_alphabet, CtcdecoderAlphabet
+from utils.pipelines import BlockedSeqPipelineStage
+
+
+class CtcDecoderSeqPipelineStage(BlockedSeqPipelineStage):
+    def __init__(self, lm=None, profile=None, beam_width=500, max_candidates=None,
+            online=False):
+        assert profile is not None, "profile argument must be provided"
+        self.p = deepcopy(profile)
+        self.beam_width = beam_width
+        self.max_candidates = max_candidates
+        self.online = online
+
+        alphabet = self.p['alphabet']
+        if alphabet is None:
+            self.alphabet = get_default_alphabet()
+        elif isinstance(alphabet, str):
+            self.alphabet = load_alphabet(alphabet)  # shall not include <blank> token
+        else:
+            self.alphabet = alphabet  # list-like
+
+        alphabet_decoder = CtcdecoderAlphabet(self.alphabet)
+        self._decoder = SeqCtcLmDecoder(self.alphabet, self.beam_width, max_candidates=max_candidates,
+            scorer_lm_fname=lm, alpha=self.p['alpha'], beta=self.p['beta'],
+            text_decoder=lambda symbols: alphabet_decoder.decode(symbols))
```

Comment:
It can, but this way seemed easier to me, because `SeqCtcLmDecoder` already loops over candidates.

Reply:
To me
> `SeqCtcLmDecoder` already loops over candidates

is actually the argument supporting the idea that `SeqCtcLmDecoder` should map the symbols on its own.

> this way seemed easier

It gets more complicated: `CtcDecoderSeqPipelineStage` needs to know an additional detail about `SeqCtcLmDecoder` and provide `text_decoder` in a way `SeqCtcLmDecoder` expects.

Reply:
Oh, I misunderstood the question in the last answer.

`text_decoder` is passed because users may want no postprocessing or some other kind of post-processing to the strings: for example, 1) UTF-8 decoding for Chinese or 2) some may need raw symbol indices as numbers, which also comes in two versions: 2.1) as views into one big numpy array, or 2.2) as separate copies of these subarrays.

Reply:
`CtcDecoderSeqPipelineStage` doesn't have an argument similar to `text_decoder`. Thus the argument 
> users may want no postprocessing or some other kind of post-processing to the strings

conflicts with current implementation. (I'm not encouraging you to introduce such arg for `CtcDecoderSeqPipelineStage` :) )

1) Are you sure C++ backend is fine with UTF-8?
2) It seems you can completely split `SeqCtcLmDecoder` and `text_decoder`: calling them sequentially: `text_decoder(self._decoder.decode())`
It conflicts with
> this way seemed easier to me, because `SeqCtcLmDecoder` already loops over candidates.

But such approach should split scopes in a cleaner way.

Reply:
Again I'm not encouraging you to follow 2. :) Just saying there should be put much more thoughts to actually make the code reusable.

Reply:
OK, removed `text_decoder` callback. And in addition removed `CtcdecoderAlphabet` class -- it was redundant.